AG-9015 - Error bars QA feedback

olegat opened the pull request:
https://ag-grid.atlassian.net/browse/AG-9015?focusedCommentId=53576

# Issues addressed:

1. Axis domain is extended to accommodate error bar values by providing additional properties to `DataController.request` using the new `''processData-prerequest'` series event.
2. Error bars are drawn in front of the markers/bars by increasing the `ZINDEX`.
3. Fixed console warnings because of misuses of `@Validate`
4. Error bars are hidden when the series is hidden by listening to the new `'visibility-changed'` events.

# Caveats

- Error bars are not drawn above the highlights.
- Axis domain is not extended for bar series (only works on line and scatter at the moment).

olegat committed:
AG-9015 Define SeriesGroupZIndexSubOrderType

This is the type of the first parameter of getGroupZIndexSubOrder. This
type is being redefined in several places and an 'error-bars' sub-group
is needed.

This change renders the error bars over the markers & bars, but not over
highlighted markers & bars (this needs to be fixed).

olegat committed:
AG-9015 Fix incorrect @Validate uses in errorBar.ts

olegat committed:
AG-9015 Hide error bars when series is hidden

This generalises the CartesianSeries.dataModelListeners property, which
has been moved / renamed to Series.seriesListeners. The property is kept
private so that the base class Series can check that the event types of
the handlers and dispatchers match. Use Series.dispatch to dispatch
events to listeners.

This is currently only tested & used by errorBar.ts.

Note: For legacy reasons, Series extends Observable. This means it now
has both addEventListener (from Observable) and addListener (from
Listeners). The former is preferred for new code.

olegat committed:
Remove dead code

olegat committed:
AG-9015 Extend the domain of line series to fit error bars

This adds a next series event 'processData-prerequest'. With this event,
the module can add, remove or modify any of the properties of the data
model before it is processed (in this case, the error bars adds the
lower & upper keys of the X & Y axes).

Only tested on line series. Probably won't work on other series.

olegat committed:
AG-9015 Extend the domain of scatter series to fit error bars

olegat committed:
AG-9015 Rename series event 'data-model' to 'data-update'

olegat committed:
AG-9015 Fix circular dependency

seresLayerManager.ts should not import series.ts

olegat commented on packages/ag-charts-community/src/chart/layers.ts:
```diff
@@ -7,6 +7,7 @@ export enum Layers {
     AXIS_ZINDEX = 20,
     SERIES_CROSSLINE_RANGE_ZINDEX = 30,
     SERIES_LAYER_ZINDEX = 500,
+    SERIES_ERRORBAR_ZINDEX = 600,
```
It's unclear whether this is necessary or whether we could just use `SERIES_LAYER_ZINDEX` instead

alantreadway replied:
🤔 I'd need to look at the live scene-graph to work that out - let's come back to it.

alantreadway commented on packages/ag-charts-community/src/chart/series/seriesEvents.ts:
```diff
@@ -0,0 +1,21 @@
+}
+
+export interface SeriesVisibilityEvent extends BaseSeriesEvent<'visibility-changed'> {
+    itemId: any;
```
Can we add a stricter typing here?

olegat replied:
I would like to, but this interface needs to be compatible with the parameters of [Series.toggleSeriesItem](https://github.com/ag-grid/ag-charts/blob/ce7c48ca175ae05309f50e8add6761b0b6f7f499/packages/ag-charts-community/src/chart/series/series.ts#L785).

I'm not sure what to use instead. Open to suggestions.

alantreadway commented on packages/ag-charts-community/src/chart/series/seriesEvents.ts:
```diff
@@ -0,0 +1,21 @@
+}
+
+export interface SeriesPrerequestDataEvent extends BaseSeriesEvent<'processData-prerequest'> {
+    props: PropertyDefinition<string>[];
```
Just wondering if `string` is the right generic type here - maybe `unknown` would be better if feasible?

olegat replied:
I'll try that.

olegat replied:
`unknown` works if I update errorBar.ts to reflect this. Done.

alantreadway commented on packages/ag-charts-enterprise/src/features/error-bar/errorBar.ts:
```diff
@@ -4,23 +4,34 @@ import type { ErrorBarPoints, ErrorBarNodeProperties } from './errorBarNode';
-    @Validate(BOOLEAN)
-    visible?: boolean = undefined;
+    @Validate(OPT_BOOLEAN)
+    visible?: boolean;
```
Q: Does this work without the `= undefined` - i.e. can a user specify `visible` without an options binding error?

olegat replied:
Oddly enough yes (https://plnkr.co/edit/m2TZ2vDoTYoHO6ra?open=main.js), although I'm unsure why. IIRC from another conversation that we had, this property shouldn't exist unless its set to something 🤷🏻‍♂️

olegat replied:
I've rebased & amended the commit that removes the explicit `= undefined` just to be safe.

alantreadway commented on packages/ag-charts-enterprise/src/features/error-bar/errorBar.ts:
```diff
@@ -88,14 +99,52 @@ export class ErrorBars
+            this.cartesianSeries.addListener('visibility-changed', (event: _ModuleSupport.SeriesVisibilityEvent) =>
+                this.onToggleSeriesItem(event)
+            )
+        );
```
Nit: Maybe just condense this to a single `.push()` call with multiple arguments?

olegat replied:
Done.

alantreadway commented on packages/ag-charts-enterprise/src/features/error-bar/errorBar.ts:
```diff
@@ -88,14 +99,52 @@ export class ErrorBars
+                valueProperty(cartesianSeries, xLowerKey, isContinuousX, { id: 'xValue' }),
+                valueProperty(cartesianSeries, xUpperKey, isContinuousX, { id: 'xValue' })
+            );
+        }
```
I would prefer if we treated the event as immutable, and instead returned additional props to append to the data-model properties, as it's a bit easier to reason about?

olegat replied:
I agree; I thought about this too initially. It's will be a bit trickier to implement because currently `Listeners.dispatch` returns `void`, but I'll see whether it's feasible to change this.

olegat replied:
Resolved in [5e1b3d780](https://github.com/ag-grid/ag-charts/commit/5e1b3d780a7dc1be647bc7937d2085bd99ca1424)

olegat commented on packages/ag-charts-enterprise/src/features/error-bar/errorBar.ts:
```diff
@@ -88,14 +99,52 @@ export class ErrorBars
+        const { cartesianSeries, xLowerKey, xUpperKey, yLowerKey, yUpperKey } = this;
+        const { isContinuousX, isContinuousY } = event;
+        event.props.push(
+            valueProperty(cartesianSeries, yLowerKey, isContinuousY, { id: 'yValue' }),
```
Note: I think that I better solution would be to use keys like `yValueLower` and add another series event for when `Series.getDomain()` is called so that `ErrorBars` can add the error values to the domain.

That would be a better solution because right now `ErrorBars` replies on private/internal implementation of the series. This solution would allow the `ErrorBars` class to function independently of the underlying series implementation (including bar series, which doesn't work at the moment).

I plan on doing this in a follow-up PR.

olegat committed:
AG-9015 Use a single push() call

olegat committed:
AG-9015 Use unknown type in SeriesPrerequestDataEvent.props

olegat committed:
AG-9015 Allow Listeners.dispatch to return values

This means the ErrorBars.onPrerequestData or longer needs to modify the
input event data (which is more intuitive).

olegat commented on packages/ag-charts-community/src/chart/series/cartesian/lineSeries.ts:
```diff
@@ -149,6 +149,11 @@ export class LineSeries extends CartesianSeries<LineContext> {
             props.push(diff(this.processedData));
         }
 
+        const listenerProps: (typeof props)[] =
```
Note: There's quite a lot of existing & new code duplication in _lineSeries.ts_ and _scatterSeries.ts_. I've filed [AG-9434](https://ag-grid.atlassian.net/browse/AG-9434) this morning for this.

alantreadway reviewed (approved):
LGTM
